infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesWorkspaceServiceAccount.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.namespace;
+
+import io.fabric8.kubernetes.api.model.rbac.KubernetesPolicyRuleBuilder;
+import io.fabric8.kubernetes.api.model.rbac.KubernetesRole;
+import io.fabric8.kubernetes.api.model.rbac.KubernetesRoleBinding;
+import io.fabric8.kubernetes.api.model.rbac.KubernetesRoleBindingBuilder;
+import io.fabric8.kubernetes.api.model.rbac.KubernetesRoleBuilder;
+import io.fabric8.kubernetes.api.model.rbac.KubernetesSubjectBuilder;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+
+/**
+ * Holds logic for preparing workspace service account.
+ *
+ * <p>It checks that required service account, roles and role bindings exist and creates if needed.
+ *
+ * @author Sergii Leshchenko
+ */
+class KubernetesWorkspaceServiceAccount {
+
+  private final String namespace;
+  private final String serviceAccountName;
+  private final KubernetesClientFactory clientFactory;
+  private final String workspaceId;
+
+  KubernetesWorkspaceServiceAccount(
+      String workspaceId,
+      String namespace,
+      String serviceAccountName,
+      KubernetesClientFactory clientFactory) {
+    this.workspaceId = workspaceId;
+    this.namespace = namespace;
+    this.serviceAccountName = serviceAccountName;
+    this.clientFactory = clientFactory;
+  }
+
+  /**
+   * Make sure that workspace service account exists and has `view` and `exec` role bindings.
+   *
+   * <p>Note that `view` role is used from cluster scope and `exec` role is created in the current
+   * namespace if does not exit.
+   *
+   * @throws InfrastructureException when any exception occurred
+   */
+  void prepare() throws InfrastructureException {
+    KubernetesClient k8sClient = clientFactory.create(workspaceId);
+
+    if (k8sClient.serviceAccounts().inNamespace(namespace).withName(serviceAccountName).get()
+        == null) {
+      createWorkspaceServiceAccount(k8sClient);
+    }
+
+    String execRoleName = "exec";
```

Comment:
Would be great to have unit tests for this class

Reply:
Agree, but mocking of fabric8-client is full of pain =)
I'll try to write some reasonable tests but don't promise

Reply:
I tried but test class contains pretty the same code that tested class does. I don't think that such test will be useful. 
I'm open for proposals, If you have an idea how to write reasonable tests for this case - you're welcome.